File: 2024/ej2/main.js
// Find problem at: https://adventofcode.com/2024/day/2
// This script should be run witho node v22.9.0
//
// An `input.txt` file is required in the same folder the solution is in
//
// The format is a newline separated list of space separated int unsigned 
// values

const fs = require("fs");
const path = require("path");


const fileInfo = fs.readFileSync(path.join(__dirname, "input.txt")).toString();

let safe = 0;
for(let line of fileInfo.split("\n")) {
  const parsedLine = line.split(" ").map((v) => parseInt(v));
  let isAscending = parsedLine[0] < parsedLine[1];
  let isValid = Math.abs(parsedLine[0] - parsedLine[1]) <= 3 && parsedLine[0] != parsedLine[1];
  for(let i = 2; i < parsedLine.length && isValid; i++) {
    isValid = Math.abs(parsedLine[i] - parsedLine[i-1]) <= 3 &&
              (
                (isAscending && parsedLine[i-1] < parsedLine[i]) ||
                (!isAscending && parsedLine[i-1] > parsedLine[i])
              );
  }
  if(isValid) {
    safe = safe + 1;
  }
}

console.log("Safe reports: ", safe)

File: 2024/ej2/main.cpp
// Todays problem was solved using c++, with g++ v13.2.0
//
// An `input.txt` file is required in the same folder the solution is runned

#include <iostream>
#include <fstream>
#include <vector>
#include <string>

enum Order {
  ASCENDING,
  DESCENDING,
  UNKNOWN
};

bool isValid(int prev, int curr, enum Order order) {
  if(prev == curr) return false;

  if(order == ASCENDING) {
    if(prev > curr) return false;
    else if(curr - prev > 3) return false;
  } else {
    if(prev < curr) return false;
    else if(prev - curr > 3) return false;
  }

  return true;
}

int main() {
  std::ifstream input_file("input.txt");

  int safe = 0;
  std::string line;
  while(getline(input_file, line)) {
    int prev = 0, curr = 0;
    bool isFirst = true, report_valid = true;
    enum Order order = UNKNOWN;

    for(auto it = line.begin(); it != line.end() && report_valid; it++) {
      char c = *it;
      if(c != ' ') {
        curr = curr * 10 + (c - '0');
      } else {
        if(!isFirst) {
          if(order == UNKNOWN) order = prev < curr ? ASCENDING : DESCENDING;
          report_valid = isValid(prev, curr, order);
        }
        isFirst = false;
        prev = curr;
        curr = 0;
      }
    }
    if(report_valid) report_valid = isValid(prev, curr, order);

    if(report_valid) {
      safe++;
    }
  }
  input_file.close();

  std::cout << "Safe reports: " << safe << std::endl;
}
